### PR TITLE
Added multiprocessing of column profiles (40% runtime reduction)

### PR DESCRIPTION
**Pull request by lettergram**
For comparison, see the two runs below (ran `United_States_Drought_Monitor__2000-2016.csv`)

I think there's still plenty of room to improve, but this reduces runtime by about 40% on this file. 

-------------------------

For `United_States_Drought_Monitor__2000-2016.csv`

Multiprocess:
```
50419680 function calls (50392635 primitive calls) in 122.032 seconds

   Ordered by: internal time
   List reduced from 1108 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      312   64.649    0.207   64.649    0.207 {method 'acquire' of '_thread.lock' objects}
       42   18.371    0.437   20.995    0.500 utils.py:53(shuffle_in_chunks)
  597/489   11.262    0.019   11.262    0.023 {built-in method numpy.array}
 16716443    3.991    0.000    3.991    0.000 {method 'search' of 're.Pattern' objects}
       14    3.897    0.278    3.897    0.278 {built-in method builtins.sorted}
       18    3.776    0.210    3.782    0.210 {pandas._libs.lib.map_infer}
        6    3.351    0.558   51.707    8.618 profile_builder.py:259(get_base_props_and_clean_null_params)
```

Single Process
```
189132806 function calls (188964701 primitive calls) in 191.153 seconds

   Ordered by: internal time
   List reduced from 1414 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      126   50.061    0.397   52.994    0.421 numerical_column_stats.py:239(_total_histogram_bin_variance)
       42   18.545    0.442   21.149    0.504 utils.py:53(shuffle_in_chunks)
10513/9311   14.103    0.001   14.114    0.002 {built-in method numpy.array}
        6   11.487    1.914   16.918    2.820 float_column_profile.py:97(_get_float_precision)
 16716414   10.113    0.000   21.751    0.000 base_column_profilers.py:47(_combine_unique_sets)
```
------------------------------

For `diamonds.csv` (no labeler)

Multiprocess:
```
2013326 function calls (1972475 primitive calls) in 5.615 seconds

   Ordered by: internal time
   List reduced from 1108 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      492    3.943    0.008    3.943    0.008 {method 'acquire' of '_thread.lock' objects}
       60    0.475    0.008    0.559    0.009 utils.py:53(shuffle_in_chunks)
```

Single process 
```
7273657 function calls (7192892 primitive calls) in 7.827 seconds

   Ordered by: internal time
   List reduced from 1473 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      304    1.945    0.006    1.945    0.006 {method 'acquire' of '_thread.lock' objects}
    23976    0.640    0.000    0.794    0.000 numerical_column_stats.py:386(_get_percentile)
       60    0.470    0.008    0.554    0.009 utils.py:53(shuffle_in_chunks)

```

### CODE REVIEW

File: dataprofiler/tests/profilers/test_base_column_profilers.py
Comment (lettergram) [inline dataprofiler/tests/profilers/test_base_column_profilers.py:277]: Can't keep a count of something which isn't called (mock the profiler) and mock's can't be pickled so this errors in multiprocess mode.

File: dataprofiler/tests/profilers/test_base_column_profilers.py
Comment (lettergram) [inline dataprofiler/tests/profilers/test_base_column_profilers.py:289]: Can't keep a count of something which isn't called (mock the profiler) and mock's can't be pickled so this errors in multiprocess mode.

File: dataprofiler/tests/profilers/test_profiler_options.py
Comment (AnhTruong) [inline dataprofiler/tests/profilers/test_profiler_options.py:128]: so, this test doesn't work when the multiprocess is enabled?

File: dataprofiler/tests/profilers/test_profiler_options.py
Comment (AnhTruong) [inline dataprofiler/tests/profilers/test_profiler_options.py:307]: same as above

File: dataprofiler/tests/profilers/test_profiler_options.py
Comment (AnhTruong) [reply dataprofiler/tests/profilers/test_profiler_options.py:128]: it that's true can we set default value for multiprocess as False?

File: dataprofiler/profilers/int_column_profile.py
Comment (AnhTruong) [inline dataprofiler/profilers/int_column_profile.py:2]: do we need it?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (AnhTruong) [inline dataprofiler/profilers/column_profile_compilers.py:60]: remove?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (AnhTruong) [inline dataprofiler/profilers/column_profile_compilers.py:127]: can we add those warning messages to the multiprocess tasks?

**Review by AnhTruong — COMMENTED** [review]
some comments. Do we have a wait to test if multiprocess is working?

File: dataprofiler/tests/profilers/test_profiler_options.py
Comment (lettergram) [reply dataprofiler/tests/profilers/test_profiler_options.py:128]: Multiprocess is a Python standard library. I don't see a reason to set it to false as, if you have anything >3 cores at the moment this should be faster.

Regarding why this test is single threaded: I set it to false because as it's a sub process the test cannot evaluate the number of times a function is executed (it doesn't own the process). You have to "own" the process for this library to see how often a given function is executed.

File: dataprofiler/profilers/int_column_profile.py
Comment (lettergram) [reply dataprofiler/profilers/int_column_profile.py:2]: No, I should remove.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (lettergram) [reply dataprofiler/profilers/column_profile_compilers.py:60]: Indeed

File: dataprofiler/profilers/column_profile_compilers.py
Comment (lettergram) [reply dataprofiler/profilers/column_profile_compilers.py:127]: I'm unclear on what you mean.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:91]: not asking for change, but i'm not sure another flag was necessary, instead, the exception could have just set 
`multiprocess_flag = False` and continue to use that, no?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:84]: does this mean if you don't have tensorflow installed you have to run as a single processes?

Comment (JGSweets) [reply dataprofiler/profilers/column_profile_compilers.py:84]: actually the call above on 72 is the one that would technically error.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:73]: this has been removed from the try statement which is a change in functionality and removes the protection for profiles. I'm not sure we want that change.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:98]: is the object changing when it goes into and comes out of multiprocessing? This is confusing to me that we have to validate we get an object out and now have to return self for every update which it should return none be default.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:169]: why do the compiler updates need self as well? is that for consistency?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (ChrisWallace2020) [reply dataprofiler/profilers/column_profile_compilers.py:127]: I think he means that currently the warnings only appear for tasks that are single process, and he wants to add them for when you iterate through the multiprocess ones, but @AnhTruong if you notice, when an exception occurs in multiprocess, the task gets added to single process, so in those cases if something truly were wrong it would go from multi_process_dict -> single_process_list -> warning, so I think its covered.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:175]: instead in the base could technically do:
```python
def __repr__(self):
    return self.__class__.__name__
```

would reduce some duplicate code

File: dataprofiler/profilers/float_column_profile.py
Comment (JGSweets) [inline dataprofiler/profilers/float_column_profile.py:198]: good move

File: dataprofiler/tests/profilers/test_categorical_column_profile.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_categorical_column_profile.py:126]: why the added spaces?

**Review by JGSweets — CHANGES_REQUESTED** [review]
I see a discussion on speed, but what are the implications on memory. last we talked, we thought it was possible that this could effectively duplicate the memory per process.

Do you have info on that?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (lettergram) [reply dataprofiler/profilers/column_profile_compilers.py:169]: Next stage, but yes consistency. Goal was all updates return (no harm in doing so)

**Review by JGSweets — COMMENTED** [review]
@lettergram mentioned that the size shouldn't matter as it is only ran on a single col. I am concerned that this will be a problem with large files even if it is a single column, but I will not block the PR.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (JGSweets) [reply dataprofiler/profilers/column_profile_compilers.py:127]: I think the intent is that if multiprocess fails, it switches over to single anyways and if that fails will get this error so @lettergram  didn't feel the need to duplicate it if it was just a fault of multiprocessing failing, which I think is fair.

Comment (JGSweets) [reply dataprofiler/profilers/column_profile_compilers.py:127]: I believe @ChrisWallace2020 is right in his assessment: `multi_process_dict -> single_process_list -> warning`

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:80]: Why is this called f?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:97]: Why have this outside of the other for loop? Is it to wait for the process to finish and validate after the thread has finished? If so, shouldnt this be after  pool.close() and pool.join()?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:87]: Hmmm, is there a common case where a certain column is often ran as a single process?

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:56]: This creates the pool workers even if the flag is set to false, I think that might not be proper.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [reply dataprofiler/profilers/column_profile_compilers.py:91]: I would like that

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:111]: This is supposed to cover the creation of the profile more than the update, so this warning isnt operating properly anymore

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:109]: Maybe just name it profile_type or col_profile_type or something

File: dataprofiler/profilers/column_profile_compilers.py
Comment (gme5078) [inline dataprofiler/profilers/column_profile_compilers.py:94]: Why would this fail and need a try-catch?

**Review by gme5078 — CHANGES_REQUESTED** [review]

File: dataprofiler/profilers/column_profile_compilers.py
Comment (AnhTruong) [reply dataprofiler/profilers/column_profile_compilers.py:127]: Ok, makes sense

File: dataprofiler/profilers/column_profile_compilers.py
Comment (AnhTruong) [reply dataprofiler/profilers/column_profile_compilers.py:91]: I'd prefer keeping single_process_flag as even in the multiprocessing mode, if we have errors, we still need to run the tasks in the single_process_list. It looks more clear to me with that extra flag.

File: dataprofiler/profilers/column_profile_compilers.py
Comment (lettergram) [reply dataprofiler/profilers/column_profile_compilers.py:84]: Yes, you do - that will always be an issue with TF, but that's why we have a seperate "single-process" operation section

File: dataprofiler/profilers/column_profile_compilers.py
Comment (lettergram) [reply dataprofiler/profilers/column_profile_compilers.py:94]: f.get() returns the error from tensorflow, if it exists.

File: dataprofiler/profilers/profiler_options.py
Comment (gme5078) [inline dataprofiler/profilers/profiler_options.py:588]: The ease at which this is added to options pleases me greatly 👍

**Review by AnhTruong — COMMENTED** [review]
lgtm now, but it'd be nice if we have the test for multiprocess

**Review by AnhTruong — APPROVED** [review]

**Review by gme5078 — APPROVED** [review]

**Pull request merged**
